Add well potential output to MSW

totto82 opened the pull request:
The common part is moved to the well interface class

totto82 commented:
jenkins build this please

totto82 commented:
The failure is expected since non-trivial well potentials are outputted.

totto82 commented:
jenkins build this update_data please

totto82 commented:
jenkins build this opm-tests=110 please

totto82 commented:
Wrong results for the field model. Needs more investigation.

totto82 commented:
I had to add some inner-iterations for the MSW model to account for change in the friction. The results are ok. But some convergence problems due to poor initial guess. Since I am not very familiar with the MSW code, I would vary much like some feedback before I finalize the code.  

I also added output for disgas and vapoil while working on the code. It should be trivial.

GitPaean commented:
> I had to add some inner-iterations for the MSW model to account for change in the friction. 

I think iterating until converged is not only for the friction. Different from Standard Well model, the density is implicit with MSW. As a result, iteration is always required. 

> I would vary much like some feedback before I finalize the code.
I will try to provide some comments soon.

totto82 commented:
Now also the density in the standard wells are recomputed. 

@GitPaean Great that you will have a look. Maybe we could have a short hangout when you have look at the code. Also there is still some convergence issues when the well potential are calculated for MSW that results in nonphysical potential values. Any hint how to improve it?

GitPaean commented:
> Also there is still some convergence issues when the well potential are calculated for MSW that results in nonphysical potential values. Any hint how to improve it?

What is the BHP values used to calculate the well potentials? If the BHP value is outside of the range of the PVT tables, it is hard for MSW. It is possible there is some assumption when coming to calculation of well potentials for MSW.

totto82 commented:
> What is the BHP values used to calculate the well potentials?

Its defaulted i.e. 1 bar. So yes it may very well be outside the pvt range.

totto82 commented on opm/autodiff/WellInterface_impl.hpp:
```diff
@@ -1171,8 +1327,9 @@ namespace Opm
     WellInterface<TypeTag>::
-    solveWellEqUntilConverged(Simulator& ebosSimulator,
+    solveWellEqUntilConverged(const Simulator& ebosSimulator,
                               const std::vector<double>& B_avg,
```
Why was the simulator non-const here and several other places in the code?

GitPaean replied:
I think it is because the function `assembleWellEq` might change the Jacobian, which is part of `ebosSimulator`? I did not re-read the code, excuse me it is wrong.

totto82 replied:
That is no-longer the case since the Jacobian part is a separate step. So then it should be changed to const.

GitPaean replied:
okay. If it compiles with `const` there, then `const` should be there.

totto82 commented on opm/autodiff/MultisegmentWell.hpp:
```diff
@@ -115,13 +115,14 @@ namespace Opm
                                     const std::vector<Scalar>& B_avg,
                                     const double dt,
+                                    const double strictBHP,
                                     WellState& well_state,
```
Passing strictBHP like this is quite ugly. Any idea how to improve this?

GitPaean replied:
We will discuss the interface things later.

totto82 commented on opm/autodiff/StandardWellV_impl.hpp:
```diff
@@ -2293,9 +2293,9 @@ namespace Opm
     computeWellRatesWithBhp(const Simulator& ebosSimulator,
-                            const EvalWell& bhp,
+                            const double& bhp,
                             std::vector<double>& well_flux,
```
Any reasons for using EvalWell here?

GitPaean replied:
Probably because we can not convert a `Scalar` to `DynamiceEvaluation`?

GitPaean commented:
> Its defaulted i.e. 1 bar. So yes it may very well be outside the pvt range.

It is not recommended in practice, while we should always be able to handle it in theory.

GitPaean commented:
The potential calculation under THP limit in the `StandardWell` model is problematic, it remains to be changed to be more advanced techniques, either based on intersection of IPR and VFP curves or a full newton solution. 

Same applies to MSW when coming to well potential calculation under THP limit. 

For the well potential calculation under BHP limit, it is relatively easy to handle. I need a little more time to think through it.

GitPaean commented:
> Part of the convergence issue was cased by the strict restriction on pressure change.
(the default max-pressure-change-ms-wells ). Is this restriction important for other models?

I believe the value we are using is pretty relaxed.

totto82 commented:
For the well potential calculation the pressure can change from say 100bars to 1bar, then the current default limitation is problematic.

GitPaean commented:
> current default limitation

Do you mean `--max-pressure-change-ms-wells`? 
Yes, it can be problematic. These parameters needs to be optimized further. I will finish the formulation part soon, then I will try to optimize these parameters.  They can be a little out-dated because the current formulation (code) is much more improved in term of convergence behavior and these parameters were set relatively conservatively to get converged results. 

And also, if the BHP changes so drastically, maybe we should update the WellState so that the initial guess is better. 

We can have a short hangout after May 1st. With this PR, are you trying to complete the well potential calculation for both StandardWell and Multisegment Well, and for both THP and BHP limits, or there is some parts are of higher priority than other parts?

totto82 commented:
> With this PR, are you trying to complete the well potential calculation for both StandardWell and Multisegment Well, and for both THP and BHP limits, or there is some parts are of higher priority than other parts?

I haven't looked at the THP part for the MSW model. The focus is on output.

GitPaean commented:
> I haven't looked at the THP part for the MSW model. The focus is on output.

okay. If I understand correctly, the key part is still to get the correct values.  Let us have a talk after May 1st.

GitPaean commented on opm/autodiff/MultisegmentWell_impl.hpp:
```diff
@@ -738,6 +840,8 @@ namespace Opm
 
             cell_perforation_pressure_diffs_[perf] = gravity_ * average_density * cell_perforation_depth_diffs_[perf];
+
+            Base::perf_densities_[perf] = average_density;
```
I do not think MSW should use `perf_densities_`. It should stay within `StandardWell`.  MSW has different way in handling the fluid densities within wellbore.

totto82 replied:
I agree that they are computed differently, but both methods computes as average perforation density.

GitPaean commented on opm/autodiff/MultisegmentWell_impl.hpp:
```diff
@@ -798,7 +902,7 @@ namespace Opm
                 const double dx_limited = sign * std::min(std::abs(dwells[seg][SPres]), relaxation_factor * max_pressure_change);
                 // const double dx_limited = sign * std::min(std::abs(dwells[seg][SPres]) * relaxation_factor, max_pressure_change);
-                primary_variables_[seg][SPres] = old_primary_variables[seg][SPres] - dx_limited;
+                primary_variables_[seg][SPres] = std::max( old_primary_variables[seg][SPres] - dx_limited, 1e5);
```
For this one, when negative BHP happens, it should switch to BHP control. Not sure we would like that. Similar code exists due to historical reason.

GitPaean commented on opm/autodiff/MultisegmentWell_impl.hpp:
```diff
@@ -638,7 +639,7 @@ namespace Opm
         // we need some intput the convergence calculations.
         // we don't need the exact values so we just use something ad-hoc
-        const std::vector<Scalar> B_avg = {800,1000,1};
+        const std::vector<Scalar> B_avg = {1000,800,1};
```
I think the values of `B` are totally off. They should be below 1?

totto82 replied:
Ops. It was meant has a quick fix for testing. I will use the competed values.

GitPaean commented:
For non-reasonable BHP limit (1 atm), I believe it is okay to report that our solution is not converged and we did not get a solution for it.

GitPaean commented:
They are just my comments when reading the PR, while I still do not have very concrete suggestion for how we should do it in a nice way. Let us continue the discussion next week.

Probably we will try to have implementation for StandardWell and MSW in a separate way first, then we try to figure out whether we can unify them. 

You do  not need to do anything to this PR at this point.

GitPaean commented:
Referring to our discussion previously. We will address only the well potential calculated based on BHP first in this PR.  I will suggest to simply through a warning within function `computeWellPotentialWithTHP()` to say we are not handling it yet now.

GitPaean commented:
For the calculation of well potentials with BHP limit, at the moment, one possible suggestion is to create a temporary `WellControls` that only contain a BHP limit. Then we solve the well equations based on this BHP limit. After we obtain the well potentials, we recover the old `WellControls`.  All these should only happen in the function `computeWellPotentials`, we re-use other functions within this function without change other functions.  

The other way is to pass a bool flag around to state we are doing well potential calculation. With this way, we might need to change more functions. 

The first option might be easier since it might potentially change less functions and can be easier to have a more unified implementation.

GitPaean commented:
> The well controls are changed in a non-intuitive way (at least for my intuition) and we really don't want the potential calculation to mess up the real well solution. I plan to test a new solution where I solve the well potential on a complete separate well container set where the WellControls only contains BHP limits.

In my opinion, since you are creating `WellControls` that only contains BHP limits, you can just do it following my first suggestion. Do it within the function `computeWellPotentials`.  You need to recover the well controls, well state and anything changed during this solution system.  To me, this can be the easiest way. 

If you go to the `BlackoilWellModel` level, then you still need to deal with each individual well. The good thing about your approach is that you intervene less with the normal running.

totto82 commented:
@GitPaean Now I have made a version that adjust the well controls instead and removed the hacking flags. Works well for the models I have tested.

totto82 commented:
jenkins build this please

totto82 commented:
jenkins build this please

GitPaean commented:
I will try to review this PR soon.  What is the status of the main motivation of the PR?

totto82 commented:
The main motivation for this PR is to compute and output the correct well potentials for MSW. It also contains an improvement for the standard wells. The PR is ready from my side. I plan a rebase to clean the commit messages, thats all, but that I will do after addressing your review.

GitPaean commented on opm/simulators/wells/WellInterface_impl.hpp:
```diff
@@ -107,6 +107,7 @@ namespace Opm
         well_efficiency_factor_ = 1.0;
 
         connectionRates_.resize(number_of_perforations_);
+        perf_densities_.resize(number_of_perforations_);
```
This should not be used in MSW at all. So it should not go to WellInterface.

totto82 replied:
I don't see why. The perf_densities are well defined also for MSW and computed in computePerfCellPressDiffs.

GitPaean replied:
But it is only temporary in that function, it is used for calculation of something else, then its life is finished there.  So it does not need to be a member at all. 

And also, they have different definition and different way of calculation. 

Unless there are some other evidence to extend its lifespan, we should not do that.

totto82 replied:
Currently, it is used in calculateBhpFromThp. If you are so strongly against this, I can revert these changes and move back the cacluateBhpFromThp code. I just tried to remove some code duplication between the standard and MSW code. Any other things you want me to change in the PR?

GitPaean replied:
Please do that. Because we should only introduce working code, remove code not working.  It is not good to introduce code not working. Depending on the plan, it might take some time before we get a chance to re-factor or correct code that not working. 

Please also remove the code (not working) in the function `computeWellPotentialWithTHP()`, instead, you can just throw a warning and return some trivial values, depending on how you want to organize the `computeWellPotentials`.  You can decide whether you want to put `computeWellPotentialWithTHP` to `StandardWell` or keep it in `WellInterface`. 

Those are the big parts with a quick review.  After that, I will provide more detailed review comments. It will be easier when we reduce the code change first.

GitPaean commented on opm/simulators/wells/WellInterface_impl.hpp:
```diff
@@ -325,6 +331,161 @@ namespace Opm
+    template<class ValueType>
+    ValueType
+    WellInterface<TypeTag>::
+    calculateBhpFromThp(const std::vector<ValueType>& rates,
```
For MSW, the top segment might not have a perforation. Unless we have some strong evidence, we should not assume to do THP related to be the same way with StandardWell.  It should be similar though. 

As we discussed, we should only address the calculation related to `BHP` limit in this PR. If we handle THP related, it will be a much more extensive PR. 

It will be good to just through a warning when handle well potential related to THP limit, including in StandardWell. 

Referring to the discussion in https://github.com/OPM/opm-simulators/issues/1607 , we should address THP related in a different PR.

totto82 replied:
I have added a warning if THP limit if the user tries to compute well potentials for cases with THP and MSW. (line 593 in MultisegmentWell_impl.hpp) 

Even though some details may be different, I would assume that most of the code will be common for MSW and standard wells regarding VFP. But since you are planning to do the VPF stuff I suggest that things gets merged as is, and you can refactor and change the code according to your need later. It is not easy to guess for me to guess what may or may not be a good code structure.

GitPaean commented on opm/simulators/wells/WellInterface_impl.hpp:
```diff
@@ -325,6 +331,161 @@ namespace Opm
+                                Opm::DeferredLogger& deferred_logger)
+    {
+        // TODO: pay attention to the situation that finally the potential is calculated based on the bhp control
+        // TODO: should we consider the bhp constraints during the iterative process?
```
Referring to the comments above. This function should thrown a warning and return a vector of zero values.  Including the existing same function in StandardWell. The implementation is outdated.

totto82 commented:
@GitPaean I have now removed the THP stuff from the common code as you suggested.

GitPaean commented:
> @GitPaean I have now removed the THP stuff from the common code as you suggested.

Thanks. You will hear from me tomorrow.

totto82 commented:
jenkins build this please

totto82 commented:
I notice that  XGRP is changed by this PR. What is XGRP?

bska commented:
> I notice that XGRP is changed by this PR. What is XGRP?

Double precision quantities per group. Typically actual and potential flow rates and cumulative totals.  Full list of supported items in [`output/eclipse/VectorItems/group.hpp`](https://github.com/OPM/opm-common/blob/4010399f08a61b5194e7ca33cb6a0f03b6bf9217/opm/output/eclipse/VectorItems/group.hpp). Since this PR modifies flow **potentials**, I'd say that's the most likely change although I'd have to look at the specific vectors to be sure.

Structure of `XGRP` is NXGRPZ (180) items for each of `NGMAXZ + 1` groups (`NGMAXZ` is item 3 of WELLDIMS, +1 for "FIELD").  FIELD results stored as last sub-vector (vector ID `NGMAXZ` in zero-based indexing).

totto82 commented:
@bska Thanks.

GitPaean commented on opm/simulators/wells/BlackoilWellModel_impl.hpp:
```diff
@@ -689,27 +698,24 @@ namespace Opm {
 
-            std::vector< Scalar > B_avg(numComponents(), Scalar() );
-            computeAverageFormationFactor(B_avg);
-
```
This potentially changes result, because it is not updated anymore for each iteration.  Do we have any document about what the correct way should be?

totto82 replied:
The B_avg does not change much during simulation. With this change it is only updated every report step i.e. when the well is created, I will be very surprised if this has any noticeable effect on the results.

GitPaean replied:
Sure. I understand it, while it probably depends on the individual case setup. 

There are some other places also changes the iteration, we will evaluate it. 

Most ideally, if this PR only output the well potentials for the cases that request the output without changing any running results, it will be much easier to review and test. 

If the iteration changes, it just adds some testing burden for the cases we have concerns with.

totto82 replied:
> There are some other places also changes the iteration, we will evaluate it.

If you send me an email with the name of the models that changes results due to this I could take a look myself.

totto82 commented:
jenkins build this please

totto82 commented on opm/autodiff/BlackoilModelParametersEbos.hpp:
```diff
@@ -72,7 +72,7 @@ SET_BOOL_PROP(FlowModelParameters, UpdateEquationsScaling, false);
 SET_SCALAR_PROP(FlowModelParameters, TolerancePressureMsWells, 0.01 *1e5);
-SET_SCALAR_PROP(FlowModelParameters, MaxPressureChangeMsWells, 2.0 *1e5);
+SET_SCALAR_PROP(FlowModelParameters, MaxPressureChangeMsWells, 1e6);
 SET_BOOL_PROP(FlowModelParameters, UseInnerIterationsMsWells, true);
```
@GitPaean This could affect performance, even thought I havn't notice any regression.

GitPaean replied:
I will test it separately. Hopefully it only does good.

GitPaean commented on opm/simulators/wells/BlackoilWellModel_impl.hpp:
```diff
@@ -880,25 +886,31 @@ namespace Opm {
         WellState well_state0 = well_state_;
 
+        if( wellsActive() )
+        {
```
Did you see some values are not initialized under some circumstances?

totto82 replied:
I did. But now the code seems to work without. I will remove the change.

GitPaean commented on opm/simulators/wells/BlackoilWellModel_impl.hpp:
```diff
@@ -415,14 +421,17 @@ namespace Opm {
         } catch ( std::runtime_error& e ) {
             const std::string msg = "A zero well potential is returned for output purposes. ";
             local_deferredLogger.warning("WELL_POTENTIAL_CALCULATION_FAILED", msg);
+        } catch (std::logic_error& e ) {
```
Where can this exception be triggered?

totto82 replied:
If the inverting of D fails it throws a numerical_error that is catched by the Deferred_logger that re-throw a logical_error in logAndCheckForExceptionsAndThrow. 

I don't know the details of the deferred_logger so I just added a fix for my problems. 
Maybe @atgeirr can explain why the exception is handled this way?

atgeirr replied:
Thanks for noticing!

The logAndCheckForExceptionsAndThrow() function does two things:
  - Writes all deferred log messages to the output rank.
  - If any rank have thrown an exception (and caught it, set the flag and passed to this function) then we throw logic_error.

I realize that this represents a loss of information compared to before for the serial case (however for the parallel case all we lost was race conditions...). Since the log messages are written, they should hopefully contain enough information to understand what happened, even if the higher level code no longer sees the exception type.

In order to re-throw the same type of exception, we would need to communicate the exception type somehow, and we have not made a facility for this. We could of course do that, it would requires us to define error codes (like in the good old days) to be used for this purpose and to change the catch spots. Perhaps adding an error code to the exception classes used also. It would be a bit of work to ensure it was done consistently across the whole system.

GitPaean replied:
It looks the `runtime_error` and `logic_error` are both related to the calculation of well potentials, I am wondering why this one is a `logic_error`? 

And since you are here, since when we terminate the simulation when we encounter `NumericalIssue`? Maybe it is just better to chop the time step to re-run to see whether the `NumericalIssue` is still there?

GitPaean commented on opm/simulators/wells/BlackoilWellModel_impl.hpp:
```diff
@@ -1067,17 +1079,75 @@ namespace Opm {
+        // Note: this must be done on all processes, even those with
+        // no wells needing testing, otherwise we will have locking.
+        std::vector< Scalar > B_avg(numComponents(), Scalar() );
+        computeAverageFormationFactor(B_avg);
```
Since `B_avg` is re-calculated here, it might not be necessary to make it a member of the `WellInterface`.  If it is seen as  a simplification of the interfaces of the original master branch, maybe it is better to do it in a separate PR so it is easier to review and evaluate?

totto82 replied:
> if it is seen as a simplification of the interfaces of the original master branch, maybe it is better to do it in a separate PR

Is that really necessary. The effect on the results seems to be below the Jenkins tolerance. Since test that don't involve output of well potentials or usage of well potential does not fail. And among the test that fails the reason is the well potential change. If you have another model you are worried about, please tell me. And I can verify that the B_avg change doesn't hurt.

GitPaean replied:
> And I can verify that the B_avg change doesn't hurt.

I understand what you mean. But on the other way, does it do any good or we have a good reason to do it?

It will result in a much smaller pull request and we do not need to evaluate the possible effects from it.

totto82 replied:
In my opinion it simplifies the interface. With the improved well potential calculations the B_avg values are needed deep down in the code. I really want to avoid passing around the B_avg values in a lot of files. With 3 wells models and both .hpp and impl_hpp files + the well interface it is significant work to change this. I think we should allow clean-up of the code while adding features. If not, I am afraid that we will never have time to improve the code.

totto82 commented:
jenkins build this please

GitPaean commented:
Jenkins failed in the following test. 
```
Report step 21/30 at day 210/300, date = 30-Jul-2015

Time step 0, stepsize 10 days.
[opm-deb:16521] *** Process received signal ***
[opm-deb:16521] Signal: Segmentation fault (11)
[opm-deb:16521] Signal code: Address not mapped (1)
[opm-deb:16521] Failing at address: 0x56514b6893a8
[opm-deb:16521] [ 0] /lib/x86_64-linux-gnu/libpthread.so.0(+0x110e0)[0x7f4a8be860e0]
[opm-deb:16521] [ 1] /var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/build-opm-simulators/bin/flow(_ZN3Opm13WellInterfaceIN5Ewoms10Properties4TTag14EclFlowProblemEE17updateWellControlERKNS1_9SimulatorIS4_EERNS_30WellStateFullyImplicitBlackoilERNS_14DeferredLoggerE+0x253)[0x5639b8b1f553]
[opm-deb:16521] [ 2] /var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/build-opm-simulators/bin/flow(_ZN3Opm13WellInterfaceIN5Ewoms10Properties4TTag14EclFlowProblemEE25solveWellEqUntilConvergedERKNS1_9SimulatorIS4_EERNS_30WellStateFullyImplicitBlackoilERNS_14DeferredLoggerE+0xb6)[0x5639b8b36806]
[opm-deb:16521] [ 3] /var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/build-opm-simulators/bin/flow(_ZN3Opm12StandardWellIN5Ewoms10Properties4TTag14EclFlowProblemEE23computeWellRatesWithBhpERKNS1_9SimulatorIS4_EERKdbRSt6vectorIdSaIdEERNS_14DeferredLoggerE+0xf0)[0x5639b8c3d230]
[opm-deb:16521] [ 4] /var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/build-opm-simulators/bin/flow(_ZN3Opm12StandardWellIN5Ewoms10Properties4TTag14EclFlowProblemEE21computeWellPotentialsERKNS1_9SimulatorIS4_EERKNS_30WellStateFullyImplicitBlackoilERSt6vectorIdSaIdEERNS_14DeferredLoggerE+0x235)[0x5639b8c3e5e5]
[opm-deb:16521] [ 5] /var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/build-opm-simulators/bin/flow(_ZN3Opm17BlackoilWellModelIN5Ewoms10Properties4TTag14EclFlowProblemEE21computeWellPotentialsERSt6vectorIdSaIdEERNS_14DeferredLoggerE+0x40e)[0x5639b8b6db8e]
[opm-deb:16521] [ 6] /var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/build-opm-simulators/bin/flow(_ZN3Opm17BlackoilWellModelIN5Ewoms10Properties4TTag14EclFlowProblemEE19prepareGroupControlERNS_14DeferredLoggerE+0x159)[0x5639b8b6e399]
[opm-deb:16521] [ 7] /var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/build-opm-simulators/bin/flow(_ZN3Opm17BlackoilWellModelIN5Ewoms10Properties4TTag14EclFlowProblemEE15prepareTimeStepERNS_14DeferredLoggerE+0x75)[0x5639b8b70715]
[opm-deb:16521] [ 8] /var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/build-opm-simulators/bin/flow(_ZN3Opm17BlackoilWellModelIN5Ewoms10Properties4TTag14EclFlowProblemEE8assembleEid+0x260)[0x5639b8b760f0]
[opm-deb:16521] [ 9] /var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/build-opm-simulators/bin/flow(+0x89fe72)[0x5639b8c4de72]
[opm-deb:16521] [10] /var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/build-opm-simulators/bin/flow(+0x8a145e)[0x5639b8c4f45e]
[opm-deb:16521] [11] /var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/build-opm-simulators/bin/flow(+0x8a575a)[0x5639b8c5375a]
[opm-deb:16521] [12] /var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/build-opm-simulators/bin/flow(_ZN3Opm12FlowMainEbosIN5Ewoms10Properties4TTag14EclFlowProblemEE12runSimulatorEv+0x18b)[0x5639b8c55c6b]
[opm-deb:16521] [13] /var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/build-opm-simulators/bin/flow(_ZN3Opm12FlowMainEbosIN5Ewoms10Properties4TTag14EclFlowProblemEE7executeEiPPc+0x233)[0x5639b8c56083]
[opm-deb:16521] [14] /var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/build-opm-simulators/bin/flow(_ZN3Opm20flowEbosBlackoilMainEiPPc+0x136)[0x5639b8a452e6]
[opm-deb:16521] [15] /var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/build-opm-simulators/bin/flow(main+0x9f8)[0x5639b8951d08]
[opm-deb:16521] [16] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf1)[0x7f4a85cfa2e1]
[opm-deb:16521] [17] /var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/build-opm-simulators/bin/flow(_start+0x2a)[0x5639b898ea2a]
[opm-deb:16521] *** End of error message ***
/var/lib/jenkins/workspace/opm-simulators-PR-builder/tests/run-regressionTest.sh: line 19: 16521 Segmentation fault      ${BINPATH}/${EXE_NAME} ${TEST_ARGS} --output-dir=${RESULT_PATH} 
```

totto82 commented:
> Jenkins failed in the following test.

Not anymore. I needed to update the currentControll in the wellState in computWellRatesWithBhp.

totto82 commented:
For reference: The simulation results on SPE9_CP_GROUP (where Jenkins claims results have changed ) is closer Ecl results.

GitPaean commented:
> For reference: The simulation results on SPE9_CP_GROUP (where Jenkins claims results have changed ) is closer Ecl results.

Good to know. Can you share some plots?

totto82 commented:
![image](https://user-images.githubusercontent.com/3223024/57850634-a79c6380-77de-11e9-94fa-42e8fdfb8311.png)

GitPaean commented:
> Random production plot from the case. For the last part of the simulation well potentials are used to distribute the group target among wells.

Thanks for sharing. 

It can be from better well potential calculation since the pressure head drop difference is updated with a solution now. 

It is also possible from the change of the way of `B_avg`, although not very likely. If it does, then it might be a good clue that it is possible we should handle `B_avg` differently.

totto82 commented:
> It is also possible from the change of the way of B_avg, although not very likely. If it does, then it might be a good clue that it is possible we should handle B_avg differently.

Of course this has nothing to do with the B_avg computation and all to do with improved well potential calculations. My point is that results changes only for cases that it should change and it is improved.

GitPaean commented on opm/simulators/wells/StandardWell_impl.hpp:
```diff
@@ -2217,6 +2246,11 @@ namespace Opm
         }
+
+
+        // rest bhp limit
```
rest -> reset

GitPaean commented on opm/simulators/wells/StandardWell_impl.hpp:
```diff
@@ -2189,12 +2189,41 @@ namespace Opm
         const int np = number_of_phases_;
         well_flux.resize(np, 0.0);
+        WellControls* wc = well_controls_;
+        const double orig_bhp = well_controls_iget_target(wc, BHP);
```
I do not understand the usage `well_controls_iget_target(wc, BHP);`. 

Please excuse me if I misread it. I think it is a wrong usage.

totto82 replied:
Based on your suggestion I changed the code such that the control is changed rather than explicitly setting the bhp in the equation. Here is just store the original bhp control target value so that I can reset the control later.

GitPaean replied:
I think `BHP` is a `enum` control mode. I do not think the usage of `well_controls_iget_target(wc, BHP)` is correct.

totto82 replied:
You are right. I havn't been working the the well controls in awhile. I guess I will have to iterate over all controls and pick the one with BHP.

GitPaean commented on opm/simulators/wells/MultisegmentWell_impl.hpp:
```diff
@@ -566,24 +566,112 @@ namespace Opm
+        const double adjustBhp = (bhp - primary_variables_[0][SPres]);
+        double adjustTotalRate = 0.0;
+        for (int p = 0; p < number_of_phases_; ++p) {
+            adjustTotalRate += copy.wellPotentials()[index_of_well_ * number_of_phases_ + p];
```
It is easy to understand how the BHP and segment pressure is adjusted. It is hard to understand how the rate is adjusted. Because the primary variable GTotal is a weighted sum of the well rates. When gas exist, this way of scaling is probably highly problematic. 

And what happens if the wellPotentials from the `copy` is not updated yet? then what happens? Please check whether the update of rates is necessary here. 

and also, since in the function, it is using `wellPoentials()` in a explicit way, maybe it is good to change the function name to be `computeWellPotentialsWithBhpLimit`, since it is not a general way of calculating well rates anymore.

totto82 replied:
> Please check whether the update of rates is necessary here.

I think it can be removed. It improved the convergence but only slightly.  I can test without it on some of the big models.

GitPaean commented on opm/simulators/wells/MultisegmentWell_impl.hpp:
```diff
@@ -566,24 +566,112 @@ namespace Opm
+        }
+
+        // compute the potential and store in the flux vector.
+        const int np = number_of_phases_;
```
I believe the following code will be equivalent to get the well rates based on `getSegmentRate(0, comp_idx)` . Basically the well rates of the top segment should be same with summing through all the perforation rates. 

Please report if they turned out to be different.

GitPaean commented:
Please revert the change related to `B_avg`. Then as a result, we will limit the change only within a few functions, and they will be only related to the calculation of well potentials. Then all the failures will be related to the output and use of well potentials. 

I am running some simulation to evaluate the parameter change related to MSW.

totto82 commented:
> Then all the failures will be related to the output and use of well potentials.

All failures are related to the well potential stuff!!!!!

GitPaean commented:
jenkins build this please

GitPaean commented:
> All failures are related to the well potential stuff!!!!!

The cases in Jenkins are simple. Many might not care about `B_avg`, but some might care. Like the following one

https://ci.opm-project.org/job/opm-simulators-PR-builder/1504/testReport/junit/(root)/mpi/compareECLFiles_flow_3D_WECON/

For bigger cases, like the cases that we care the most, mostly likely `B_avg` will make a difference in term of iteration behavoir, then we have to evaluate.  If the change of `B_avg` is not intentional (having a good reason), we better not deal with it. 

A pull request will be good if it can fulfill its purpose with making minimal change. 

I am here to help. I am not here to make problems.

GitPaean commented:
By changing the `MaxPressureChangeMsWells` to be `1e+6`, the two cases I ran, one is faster, one is slower.  More cases will be tested.

GitPaean commented:
> By changing the MaxPressureChangeMsWells to be 1e+6, the two cases I ran, one is faster, one is slower. More cases will be tested.

I conducted running of another 4 examples, 3 of them are slightly slower, while one of them is faster, however, the one faster is faster in a more significant way. 

I think it is okay to change this parameter.

totto82 commented:
> I am here to help. I am not here to make problems.

I know. Thanks!!! I will do what you suggest, but keep in mind that all changes you suggest makes me spend more time on this.

GitPaean commented:
> but keep in mind that all changes you suggest makes me spend more time on this.

Thanks. I have my concerns too. One is that I do think it cost less time than running through the important cases to evaluate possible impacts.  And also, if it causes later problems, for sure, it will be a bigger effort to fix it. It is good to limit the scope of PRs. 

I can help some patches if needed. Just let me know.

totto82 commented:
I have reverted the B_avg change. I am still looking into the WECON failure.

alfbr commented:
With this PR I am still getting messages like this:

Warning: Well potential calculation is not supported for multisegment wells 
A well potential of zero is returned for output purposes. 
If you need well potential to set the guide rate for group controled wells 
you will have to change the A-1 well to a standard well

Should not the message be removed when this PR is merged?

bska commented:
> With this PR I am still getting messages like this:
> 
>> Warning: Well potential calculation is not supported for multisegment wells

Does the message say **exactly**
```
... not supported for multisegment ...
```
or is slightly more refined as
```
... not supported for thp controlled multisegment ...
```
with the difference between the two being "thp controlled"?

If it's the first, then that's likely to be a bug.  If it's the second, then you've "just" bumped into a different limitation (cf. https://github.com/OPM/opm-simulators/blob/1e4ab75716dbcc36637496b3d8651b9fd1dcd383/opm/simulators/wells/MultisegmentWell_impl.hpp#L570-L573)

GitPaean commented on opm/simulators/wells/StandardWellV_impl.hpp:
```diff
@@ -2265,12 +2273,42 @@ namespace Opm
+        well_controls_set_current(wc, bhp_index);
+
+        // iterate to get a more accurate well density
+        if (iterate) {
```
I understand the problem better now.

I do not have access to reference result. It is possible for well potential calculation, it needs to solve the well equation first, to update the perforation pressure drops to be proper first. 

And also, at the same time, for some other purposes, it might not need it, because the perforation pressure drops are treated in an explicit manner.  There are specified perforation pressure drops there already. 

Maybe we should split them `computeWellPotentialsWithBhp` and you call `computeWellRatesWithBhp` within the previous function. 

Will this make the interface change of `B_avg` less, instead of using a boolean input arguments to put them together?

totto82 replied:
But now I have done the work. I really don't want to spend another half day changing it all back again.

GitPaean replied:
Sure. It is fine. I put it on my list. I will handle it.

totto82 commented:
> With this PR I am still getting messages like this:

The messages was accidentally reintroduced during the process of addressing code review. It is long gone now.

alfbr commented:
@totto82 are you sure, this is from git master built this morning, tagged this morning. I double checked that I had this PR in. I can triple check of course.

> Does the message say **exactly**
> 
> ```
> ... not supported for multisegment ...
> ```

@bska, yes the message is cut'n'paste.

totto82 commented:
@alfbr It was removed 3h ago.

alfbr commented:
Thanks, will give it another go.

totto82 commented:
> Thanks, will give it another go.

I suggest that you wait. After the latest change of code there is some issues with the WECON computations. I.e. a potential but in the PR in its current form.

alfbr commented:
Sure, just give me a heads-up when it is ready.

totto82 commented:
jenkins build this please

totto82 commented:
Anything more? This Pr needs an update of data. I have verified that all failures are due to improved well potential calculations.

bska commented:
> This Pr needs an update of data.

Please hold all data updates until we've merged #1848 and OPM/opm-tests#119.

GitPaean commented:
> Anything more?

Thanks for the patience. I will review it today and tomorrow. It is possible I will crate PR directly towards your branch. 

It is not easy to tell the best way to do the things related as you can see. I will need some functionality in the immediate future. So let us spend a little more efforts before it goes in. 

You do not need to do anything more without my further request.

GitPaean commented:
> Please hold all data updates until we've merged #1848 and OPM/opm-tests#119.

No problem. We will wait to do it after merging #1848.

GitPaean commented:
It is good if you can do a rebase. So it might be easier to create PR against your branch.

GitPaean commented on opm/simulators/wells/StandardWellV_impl.hpp:
```diff
@@ -2265,12 +2273,42 @@ namespace Opm
+
         const int np = number_of_phases_;
         well_flux.resize(np, 0.0);
+        WellControls* wc = well_controls_;
```
Without `iterate` to be `true`,  the setting of the `bhp` limit should not happen here.  An immediate change will be put these into the `if (iterate) {`, including the following code
```
        // reset bhp limit
        well_controls_iset_target(wc, bhp_index, orig_bhp);
        well_controls_set_current(wc, orig_current);
    }	

```
Please do not do anything now to see whether we can find a better solution.

alfbr commented:
I believe a change on this PR during the last 24 hours has hurt numerical convergence/robustness. I have a couple of cases over here that fails to run now, but successfully ran through two days ago.

totto82 commented:
@alfbr Maybe you could contact me directly with some more details.

GitPaean commented:
> I believe a change on this PR during the last 24 hours has hurt numerical convergence/robustness. I have a couple of cases over here that fails to run now, but successfully ran through two days ago.

Thanks for reporting. That is the most worrying and difficult part. Please provide more details through emails if necessary.

GitPaean commented on opm/autodiff/BlackoilModelParametersEbos.hpp:
```diff
@@ -72,7 +72,7 @@ SET_BOOL_PROP(FlowModelParameters, UpdateEquationsScaling, false);
 SET_BOOL_PROP(FlowModelParameters, UseUpdateStabilization, true);
 SET_BOOL_PROP(FlowModelParameters, MatrixAddWellContributions, false);
 SET_SCALAR_PROP(FlowModelParameters, TolerancePressureMsWells, 0.01 *1e5);
-SET_SCALAR_PROP(FlowModelParameters, MaxPressureChangeMsWells, 2.0 *1e5);
```
@alfbr , @totto82 , it is mostly likely this change makes some cases fail.

GitPaean replied:
By changing it to `2.e5`, the running is recovered, while some slightly different iteration process was shown. One more testing is ongoing and it will be reported back.

GitPaean replied:
Did not figure out why the iteration behavior changed compared with master branch. But with resetting it back to `2.e5`, the running of the obtained problematic case is recovered.

totto82 replied:
I hope #1855 will fix the problem. Testing in progress......

totto82 replied:
#1855 fixed the problem.

GitPaean replied:
Good! What is the actual problem? Some exceptions we could not catch?

totto82 replied:
With #1855 all exceptions should be caught. I have added a partial fix for this in this PR. I will remove it when #1855 is merged.

GitPaean replied:
Good! How the performance looks like? similar?

totto82 replied:
I am testing it at the moment. I.e. The same case with `2e5`

totto82 replied:
The 1e6 speeds up the computations due to less convergence issues.

GitPaean replied:
That is good to know. Is this the one problematic case we were trying to fix?  Can you post the performance information here so that I do not need to run it.

GitPaean commented on opm/simulators/wells/StandardWellV_impl.hpp:
```diff
@@ -2265,12 +2273,42 @@ namespace Opm
     StandardWellV<TypeTag>::
     computeWellRatesWithBhp(const Simulator& ebosSimulator,
-                            const EvalWell& bhp,
+                            const std::vector<Scalar>& B_avg,
```
I am trying to rewriting this comment. 

I suggest to split this function to be two functions. 

The first part basically applies to the part that when `iterate` to be `true`. It includes resetting the bhp limit at the beginning, solving the well equations and update the explicit quantities, and the recover the the well controls at the end. Please make a proper name for it. 

The second function will be the original `computeWellRatesWithBhp` unchanged. 

With this two functions, you can make a new function called `computeWellPotentialsWithBhpLimit`.

GitPaean replied:
The resetting of the BHP limit in the current form is not very proper, since it discard the fact that there might be other controls included there at the same time. It did not cause problem is because it is only used when `iterate` to be `true`. 

And also, we have two places resetting the BHP limits. The other places is the `computeWellPotentials` in `BlackoilWellModel`. We should not need to do it twice. 

I will suggest we only do it in one places.  The well control only contains a BHP limit can be created using some function as follows, 
```
    template<typename TypeTag>
    struct WellControls*
    WellInterface<TypeTag>::
    creatingWellControlWithBhpTargetOnly(const double bhp)
    {
        struct WellControls* controls = well_controls_create();
        const double invalid_alq = -1.e100;
        const double invalid_vfp = -1.;
        well_controls_add_new(BHP, bhp, invalid_alq, invalid_vfp, NULL, controls);

        assert(well_controls_get_num(controls) == 1);
        well_controls_set_current(controls, 0);
        return controls;
    }
```

Then we can do the resetting the well control in some manner as follows,
```
        // back up the old control data
        const struct WellControls* original_well_controls = well_controls_;
        const int original_current_control = well_controls_get_current(original_well_controls);
```

then 
```
well_controls_ = this->creatingWellControlWithBhpTargetOnly(bhp);
```

At the end, we recover the old control

```
        // recovering the old well control
        well_controls_destroy(well_controls_);
        well_controls_ = original_well_controls;
        well_controls_set_current(well_controls_, original_current_control);
```

All these should be the first function mentioned above.  We do not need to do it in the `computeWellPotentials` in `BlackoilWellModel`.

GitPaean replied:
In MultisegmentWell, the suggestions will be similar. 

Ideally, we should not need to recreate the `well_container`. I do see that it might needs less cleaning and recovering work by creating a new one.

GitPaean replied:
Great you find that the well potentials can be more correct when updating the explicit quantities after solving the system til converged under the desired BHP target. 

It is possible in the future we should always do that, while let us do it step by step.

GitPaean commented on opm/simulators/wells/BlackoilWellModel_impl.hpp:
```diff
@@ -544,58 +543,60 @@ namespace Opm {
-                            well_state_.setEffectiveEventsOccurred(w, false);
-                        } else {
-                            wellTestState_.openWell(well_name);
+                if (allow_closing_opening_wells) {
```
if `allow_closing_opening_wells` is `false`, does this mean, even a well is shut due to economic or physical reason, we still need to calculate the well potentials?

maybe it is possible for the wells shut due to economic reasons, while it can be hard for the wells shut due to physical reason. 

I think it is reasonable to assume that we set the well potentials to be zero or not trying to calculate the well potentials for the wells are shut.

GitPaean commented on opm/simulators/wells/MultisegmentWell_impl.hpp:
```diff
@@ -541,24 +542,73 @@ namespace Opm
+                            std::vector<double>& well_flux,
+                            Opm::DeferredLogger& deferred_logger)
+    {
+        // store a copy of the well state, we don't want to update the real well state
```
Suggestion on how to handle this function is similar to the previous comment with StandardWell. The well_controls_ is not reset here, it is reset in `BlackoilWellModel`. It might not be correct since a THP limit can be there at the same time. 

Eventually, if our code can just set the BHP and THP together, then we just solve it, the resulting well rates will be the well potentials, that will be great and that should be our goal, while I do not think we can do that in this PR, at least not for MSW.  We still have to calculate under BHP and THP separately and pick the one gives the smaller well potential value.

GitPaean commented:
My current comments are done. Please address following the comments. 

This PR is much better now compared with its original form. And it is hard to tell what is good way / correct way to do it. Continuous improvements will be required even after this PR, while it can be easier just try to little by little, which can result better understanding and smaller risk.  

Currently, due to the existence of the explicit quantities, StandardWell has more complicated logic than MSW in some sense.  Maybe we should switch to implicit for StandardWell too. Then the logic can be simpler. We just solve it, not updating a little bit to make the result slightly better. But it is good you found out that one more updating of the explicit quantities can improve the well potential calculation, which is an important clue.

GitPaean commented:
By `complicated logic` related to StandardWell, I mean we have some explicit quantities, but sometimes we want to update it by solving once, because we found it makes the result slightly better. It is possible by updating it twice, the result will be better slightly further. If we keep updating it until result not change anymore, it is kind of implicit in some sense. Maybe by treating them implicit, we will be able to simplify the logic.

alfbr commented:
A conflict resolution is highly appreciated, would love to build again and test more :-)

totto82 commented:
jenkins build this please

totto82 commented:
jenkins build this please

totto82 commented:
jenkins build this please

bska commented:
jenkins build this please

totto82 commented:
jenkins build this update_data please

totto82 commented:
jenkins build this opm-tests=110 please

totto82 commented:
jenkins build this opm-tests=110 please

bska commented:
The data update PR (OPM/opm-tests#110) was created before we merged OPM/opm-common#707 and its associate data update, OPM/opm-tests#114. Therefore it is likely that OPM/opm-tests#110 is stuck referencing older data. In previous cases like this I've been able to work around that problem by closing the data update PR&mdash;here OPM/opm-tests#110&mdash;for about one hour, before reopening it and rerunning `update_data`.

I would however advice not doing so until all reviewers of this PR has approved of the changes so that the only remaining step is the final data update.

atgeirr commented:
This PR can now be merged, when green. Thanks for your patience and the changes you have made. The data update things suggested by @bska above are ok to to now.

totto82 commented:
jenkins build this update_data please

atgeirr commented:
jenkins build this opm-tests=110 please
